astropy/wcs/src/distortion.c

```diff
@@ -79,11 +79,11 @@ image_coord_to_distortion_coord(
   assert(lookup != NULL);
   assert(axis < NAXES);
 
-  /* The "- 1./stepsize" is here because the input coordinates are 1-based,
+  /* The "- 1" is here because the input coordinates are 1-based,
      but this is a C-array underneath */
   result = (
       ((img - lookup->crval[axis]) / lookup->cdelt[axis]) +
-      lookup->crpix[axis]) - 1.0/lookup->cdelt[axis];
+      lookup->crpix[axis]) - 1.0;
```

Comment:
I've been thinking about this change, it's been a while since I looked at this code so it took me a while.
Suppose the lookup table (LT) is two times smaller than the image, i.e. cdelt=2. For simplicity let's take CRPIX=0 and CRVAL=0. Coordinates are 1-based because this is FITS. The first pixel in the image array, img_x=1, should map into the first pixel of the LT, lt_x=1. And because it's a C array this corresponds to index 0. So `img_x/cdelt + correction` should result in 0 (ignoring crpix and crval). `correction` needs to be `-1/cdelt` to get 0. This is because 1 pixel in the image corresponds to 1/cdelt pixels in the LT. 
The first two pixels in the image array map into the first pixel in the LT, and so on. img_x=3 will map into index 1, i.e. lt_x=2.
Similarly, going in the opposite direction, from LT to image, there needs to be a correction to get the C array index and in this case the correction is `-cdelt`, or `-1*cdelt`, because 1 pixel in the LT corresponds to cdelt pixels in the image. For example, if cdelt was 64, the correction going from LT to image should be -64.

I hope this makes sense. The original is correct in my opinion.
To summarize, internally all computations assume 1-based coordinates. To get an index into the array, a correction is applied and it is equal to the fractional pixel in the LT corresponding to a full pixel in the image.

Reply:
Thanks for looking at this! There are two spots where I'm not following along:

If I'm understanding right, it seems like you're applying the conversion from FITS pixels to array indices (the offset of 1) in image space and not LT space, but I think LT space is where it should happen. In this function we have a one-based img_x as input, and we want the corresponding value from the LT array, so a workflow of converting to a 1-based lt_x, and then subtracting 1 to get an array index makes sense to me. Whereas this code (before this PR) translates to `lt_array_index = (img_x - crval) / cdelt + crpix - 1/cdelt` => `lt_array_index = ([img_x - 1] - crval) / cdelt + crpix`, which I interpret as converting img_x to an array index into the image, and then converting to a LT index, but that's done with the 1-based crpix and crval, which doesn't seem right.

CRPIX and CRVAL must be 1-based coordinates, right? That's how `WCS` seems to handle it on the image<->world side. So when we set crpix=crval=0, it's not the case that img_x=1 corresponds to lt_x=1. Instead, shouldn't lt_x=1 correspond to img_x=2 (while lt_x=0.5 is img_x=1 and lt_x=0 is img_x=0)? So then `img_x/cdelt + correction` is `2/2 + correction`, and to get array index 0 corresponding to lt_x=1, we need `correction=-1`. This gives `[(img_x - crval) / cdelt + crpix] - 1`, which to me has the straightforward interpretation of converting the 1-based img_x to a 1-based lt_x with everything in the square brackets, and then converting lt_x to a 0-based array index into the LT array.

Reply:
> it seems like you're applying the conversion from FITS pixels to array indices in image space and not LT space

Yes, I think the offset needs to be applied in `input` space. Essentially the correction needs to be one input pixel expressed in output coordinates. I think words explain this better than numbers. In either direction, the first element of the image array needs to map into the first element of the LT array, and the last element of the image into the last element of the LT. In between we interpolate. 

> So when we set crpix=crval=0, it's not the case that img_x=1 corresponds to lt_x=1. Instead, shouldn't lt_x=1 correspond to img_x=2

I think not, see above. When you design the look up table and determine what the WCS keywords need to be, the starting point is that he first and last elements need to map. This determines the CRVAL and CRPIX.

Reply:
Is there a bigger story behind this? The argument here is which two pixels are used in the interpolation, with a difference of one pixel. Unless the differences in adjacent pixels are large the end differences should be negligible.

Reply:
> Is there a bigger story behind this? The argument here is which two pixels are used in the interpolation, with a difference of one pixel. Unless the differences in adjacent pixels are large the end differences should be negligible.

You're right that this should be a very small difference, but I still think it's something that astropy should get right and not just close enough. I care as part of trying to align PSP/WISPR images with as much sub-pixel precision as possible, to support a starfield estimation and removal process, so I've been focusing on small details. I must have come across this (over a year ago) while testing that I was building and using `DistortionLookupTable`s correctly.

> In either direction, the first element of the image array needs to map into the first element of the LT array, and the last element of the image into the last element of the LT

This in particular has implications for the other half of this PR, handling the LT during wcs slicing. I agree that a well-designed LT will have its first and last pixels map to the first and last image pixels, which then constrains the values of CRPIX/VAL/DELT, but the way I read the Calabretta paper laying this all out, this isn't required, and the first/last LT pixels could map to somewhere beyond the bounds of the image. They call this out on p10 as a way to handle the LT when extracting a sub-image from the main image, by just updating the LT's CRVAL, and only optionally cropping out the LT pixels that don't map anywhere in the subimage and updating CRPIX to match (as I understand those two sentences).

So I really think that the CRPIX, VAL = 0, 0 from your example means that the LT and image are "aligned" at their 0th pixels or -1st array indices (if negative indices didn't wrap around), and then the first actual LT pixel maps to the second image pixel (since cdelt=2). And I think that still implies that `correction` must be `-1`, as in my earlier comment. To have the first image pixel map to the first LT pixel, while CDELT != 1, the choice would be CRPIX, VAL = 1, 1 (or `1+n, 1+n*cdelt`), which will also require `correction=1` to produce the right array indices.

If you still don't agree, I wonder if you could look at the `test_DistortionLookupTable` I added in this PR. (I just reorganized it a bit so it was easier to double-check.) The test passes only with this PR in place---what would you change so it passes without this PR?

Again, thanks for your time on this!